Packages/com.unity.inputsystem/InputSystem/Editor/ProjectWideActions/ProjectWideActionsAsset.cs

```diff
@@ -97,10 +105,51 @@ private static InputActionAsset CreateNewActionAsset()
                     AssetDatabase.AddObjectToAsset(actionReference, asset);
                 }
             }
+        }
 
-            AssetDatabase.SaveAssets();
+        /// <summary>
+        /// Updates the input action references in the asset by updating names, removing dangling references
+        /// and adding new ones.
+        /// </summary>
+        /// <param name="asset"></param>
+        internal static void UpdateInputActionReferences()
+        {
+            var asset = GetOrCreate();
+            var existingReferences = InputActionImporter.LoadInputActionReferencesFromAsset(asset).ToList();
 
-            return asset;
+            // Check if referenced input action exists in the asset and remove the reference if it doesn't.
+            foreach (var actionReference in existingReferences)
+            {
+                var action = asset.FindAction(actionReference.action.id);
+                if (action == null)
+                {
+                    actionReference.Set(null);
+                    AssetDatabase.RemoveObjectFromAsset(actionReference);
+                }
+            }
+
+            // Check if all actions have a reference
+            foreach (var action in asset)
+            {
+                var actionReference = existingReferences.FirstOrDefault(r => r.m_ActionId == action.id.ToString());
+                // The input action doesn't have a reference, create a new one.
+                if (actionReference == null)
+                {
+                    var actionReferenceNew = ScriptableObject.CreateInstance<InputActionReference>();
+                    actionReferenceNew.Set(action);
+                    AssetDatabase.AddObjectToAsset(actionReferenceNew, asset);
+                }
+                else
+                {
+                    // Update the name of the reference if it doesn't match the action name.
+                    if (actionReference.name != InputActionReference.GetDisplayName(action))
+                    {
+                        AssetDatabase.RemoveObjectFromAsset(actionReference);
+                        actionReference.name = InputActionReference.GetDisplayName(action);
+                        AssetDatabase.AddObjectToAsset(actionReference, asset);
```

Comment:
This is making me think that InputActionReference's displayed in other parts of the UI won't be renamed as they have their own InputActionReference (with asset / action id and **cached name**)

(But perhaps when those are displayed the name is refreshed there too?)

Reply:
> (But perhaps when those are displayed the name is refreshed there too?)

From my testing, this is what currently happens.